apps/emqx_gateway_mqttsn/src/emqx_mqttsn_channel.erl

```diff
@@ -2083,42 +2063,56 @@ handle_timeout(
     retry_delivery,
     Channel = #channel{conn_state = asleep}
 ) ->
-    {ok, reset_timer(retry_timer, Channel)};
+    {ok, reset_timer(retry_delivery, Channel)};
 handle_timeout(
     _TRef,
-    retry_delivery,
-    Channel = #channel{session = Session, clientinfo = ClientInfo}
+    _Name = expire_awaiting_rel,
+    Channel = #channel{conn_state = disconnected}
 ) ->
-    case emqx_mqttsn_session:retry(ClientInfo, Session) of
-        {ok, NSession} ->
-            {ok, clean_timer(retry_timer, Channel#channel{session = NSession})};
+    {ok, Channel};
+handle_timeout(
+    _TRef,
+    Name = expire_awaiting_rel,
+    Channel = #channel{conn_state = asleep}
+) ->
+    {ok, reset_timer(Name, Channel)};
+handle_timeout(
+    _TRef,
+    Name,
+    Channel = #channel{session = Session, clientinfo = ClientInfo}
+) when Name == retry_delivery; Name == expire_awaiting_rel ->
+    case emqx_mqttsn_session:handle_timeout(ClientInfo, Name, Session) of
+        {ok, Publishes, NSession} ->
+            NChannel = Channel#channel{session = NSession},
+            handle_out(publish, Publishes, clean_timer(Name, NChannel));
         {ok, Publishes, Timeout, NSession} ->
             NChannel = Channel#channel{session = NSession},
             %% XXX: These replay messages should awaiting register acked?
-            handle_out(publish, Publishes, reset_timer(retry_timer, Timeout, NChannel))
+            handle_out(publish, Publishes, reset_timer(Name, Timeout, NChannel))
     end;
 handle_timeout(
     _TRef,
-    expire_awaiting_rel,
+    {emqx_session, _Name},
```

Comment:
Nowhere yet. 🫠 The idea is that we will use `emqx_session:ensure_timer/3` sooner than later somewhere in the persistent session implementation code, so there's some stuff to make sure this will work.

Reply:
Backed them out as part of e3b5d75, this stuff will go in the followup PR.